CoreScriptsRoot/ServerCoreScripts/ClientChat/ChannelsBar.lua

```diff
@@ -0,0 +1,322 @@
+	lockScrollChannelsFrame = true
+
+	local newPageNum = self.CurPageNum + dir
+	if (newPageNum < 0) then
+		newPageNum = 0
+	elseif (newPageNum > 0 and newPageNum + 4 > self.NumTabs) then
+		newPageNum = self.NumTabs - 4
+	end
+
+	self.CurPageNum = newPageNum
+
+
+	local tweenTime = 0.15
+	local endPos = UDim2.new(-(self.CurPageNum * 0.25), 0, 0, 0)
+
+	self.PageLeftButton.Visible = (self.CurPageNum > 0)
+	self.PageRightButton.Visible = (self.CurPageNum + 4 < self.NumTabs)
+
+	local function UnlockFunc()
+		lockScrollChannelsFrame = false
+	end
+	self.ScrollerFrame:TweenPosition(endPos, Enum.EasingDirection.InOut, Enum.EasingStyle.Quad, tweenTime, true, UnlockFunc)
+end
+
+function methods:FadeOutBackground(duration)
+	self.BackgroundTweener:Tween(duration, 1)
+end
+
+function methods:FadeInBackground(duration)
+	self.BackgroundTweener:Tween(duration, 0)
+end
+
+function methods:FadeOutText(duration)
+	self.TextTweener:Tween(duration, 1)
+end
+
+function methods:FadeInText(duration)
+	self.TextTweener:Tween(duration, 0)
+end
+
+function methods:CreateTweeners()
+	self.BackgroundTweener:CancelTween()
+	self.TextTweener:CancelTween()
+
+	self.BackgroundTweener = moduleTransparencyTweener.new()
+	self.TextTweener = moduleTransparencyTweener.new()
+
+	--// Register BackgroundTweener objects and properties
+	self.BackgroundTweener:RegisterTweenObjectProperty(self.PageLeftButton, "ImageTransparency")
+	self.BackgroundTweener:RegisterTweenObjectProperty(self.PageRightButton, "ImageTransparency")
+	self.BackgroundTweener:RegisterTweenObjectProperty(self.PageLeftButtonArrow, "ImageTransparency")
+	self.BackgroundTweener:RegisterTweenObjectProperty(self.PageRightButtonArrow, "ImageTransparency")
+
+	--// Register TextTweener objects and properties
+	
+end
+
+--///////////////////////// Constructors
+--//////////////////////////////////////
+ClassMaker.RegisterClassType("ChannelsBar", methods)
+
+function module.new()
+	local obj = {}
+
+	local BaseFrame, ScrollerFrame, PageLeftButton, PageRightButton, LeaveConfirmationFrame, LeaveConfirmationNotice = CreateGuiObject()
+	obj.GuiObject = BaseFrame
+	obj.ScrollerFrame = ScrollerFrame
+	obj.PageLeftButton = PageLeftButton
+	obj.PageRightButton = PageRightButton
+	obj.LeaveConfirmationFrame = LeaveConfirmationFrame
+	obj.LeaveConfirmationNotice = LeaveConfirmationNotice
+
+	obj.PageLeftButtonArrow = obj.PageLeftButton.ArrowLabel
+	obj.PageRightButtonArrow = obj.PageRightButton.ArrowLabel
+
+	obj.ChannelTabs = {}
+	obj.NumTabs = 0
+	obj.CurPageNum = 0
+
+	obj.BackgroundTweener = moduleTransparencyTweener.new()
+	obj.TextTweener = moduleTransparencyTweener.new()
+
+	ClassMaker.MakeClass("ChannelsBar", obj)
+
+	obj:CreateTweeners()
+
+	PageLeftButton.MouseButton1Click:connect(function() obj:ScrollChannelsFrame(-1) end)
+	PageRightButton.MouseButton1Click:connect(function() obj:ScrollChannelsFrame(1) end)
+
+	--// Have to wait to tween until it's properly parented to PlayerGui
+	spawn(function()
+		wait()
```

Comment:
Is there an event you could wait for rather than just a shotgun kind of approach? Maybe you could do some AncestryChanged:wait() here.